enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,562 @@
+are configured, will load-balance the traffic depending on how the infrastructure network is configured.
+For example, if the ToR router is a Linux machine, we can control which hash policy to use for multipath routes with
+`fib_multipath_hash_policy` in sysctl and choose if we want L3 or L4 load balancing. If a Cisco router is used, the `maximum-paths` option can be helpful. It really
+depends on the upstream router and BGP infrastructure that is in place, and not managed by OpenShift.
+
+
+### baremetal-runtimecfg
+
+[baremetal-runtimecfg](https://github.com/openshift/baremetal-runtimecfg) is used for rendering the manifests and templates of the machine-config-operator project, it also supports runtime update of the configuration templates (e.g: update FRR config in case a new BGP peer added to the BGP Configuration for the platform) based on the current system status.
+
+The next capabilities are supported by the `baremetal-runtimecfg`:
+- Based on the network configuration on the host, it'll find which BGP configuration needs to be applied on the host. For that, it'll follow that logic:
+  - If controlPlaneLoadbalancer is of the type BGP, then it'll get all speaker configurations.
+  - Each node has an `InternalIP` address that we can find in `node.Status.Addresses`. We'll have a function that find which BGP speaker subnet matches with the node IP. If more than one speaker has matched, we'll return a failure because we don't
+    expect the control plane nodes to be connected to more than one subnet used for BGP peering. For example, if the machine has the InternalIP being 192.168.0.10 and there is a speaker that has the subnet CIDR being 192.168.0.0/24, it'll match and the speaker configuration will be applied to that machine.
+- `renders templates` using [values provided at command line](https://github.com/openshift/baremetal-runtimecfg/blob/master/cmd/runtimecfg/render.go), parameters retrieved from the cluster (e.g: control plane nodes) and
+  - api-vips
+  - ingress-vips
+- `frr monitor`: verify that the desired configuration is synced with the cluster. As seen before, it's used from a side-car container in the FRR pod.
+
+### API Extensions
+
+We introduce new CRDs via https://github.com/openshift/api/pull/1321.
+
+### Implementation Details/Notes/Constraints
+
+#### Bootstrap
+
+This paragraph is WIP.
+
+how does the bootstrap node know its configuration?
+
+BGP configuration is written to kube-apiserver as manifests during installation. But... how do we
+know which subnet we're on so we know which to apply for the bootstrap node?
+
+#### Control plane cold start
+
+This paragraph is WIP.
+
+A critical difference between this solution and keepalived (and external loadbalancers), is that
+keepalived's configuration is static on disk, whereas with BGP we update the configuration by
+rendering a new `frr.conf` using values ultimately sourced from the api server. This means that
```

Comment:
Can we not do the same thing as with keepalived (which also renders its config based on values retrieved from the API, for the record)? The config file is stored on the host filesystem so that if the node goes down, when it comes back up the previous config is still there and keepalived is able to stand back up based on the previous state.

Reply:
The config file is stored on the filesystem, but I'll double check.